x/exp/batch/batch.go

```diff
@@ -0,0 +1,387 @@
+	"github.com/cedar-policy/cedar-go"
+	"github.com/cedar-policy/cedar-go/internal/ast"
+	"github.com/cedar-policy/cedar-go/internal/consts"
+	"github.com/cedar-policy/cedar-go/internal/eval"
+	"github.com/cedar-policy/cedar-go/types"
+)
+
+// Ignore returns a value that should be ignored during batch evaluation.
+func Ignore() types.Value { return eval.Ignore() }
+
+// Variable returns a named variable that is populated during batch evaluation.
+func Variable(name types.String) types.Value { return eval.Variable(name) }
+
+// Request defines the PARC and map of Variables to batch evaluate.
+type Request struct {
+	Principal types.Value
+	Action    types.Value
+	Resource  types.Value
+	Context   types.Value
+	Variables Variables
+}
+
+// Variables is a map of String to slice of Value.
+type Variables map[types.String][]types.Value
+
+// Values is a map of String to Value.  This structure is part of the result and
+// reveals the current variable substitutions.
+type Values map[types.String]types.Value
+
+// Result is the result of a single batched authorization.  It includes a
+// specific Request, the Values that were substituted, and the resulting
+// Decision and Diagnostics.
+type Result struct {
+	Request    types.Request
+	Values     Values
+	Decision   types.Decision
+	Diagnostic types.Diagnostic
+}
+
+// Callback is a function that is called for each single batch authorization with
+// a Result.
+type Callback func(Result)
+
+type idEvaler struct {
+	PolicyID types.PolicyID
+	Policy   *ast.Policy
+	Evaler   eval.BoolEvaler
+}
+
+type idPolicy struct {
+	PolicyID types.PolicyID
+	Policy   *ast.Policy
+}
+
+type batchEvaler struct {
+	Variables []variableItem
+	Values    Values
+
+	policies []idPolicy
+	compiled bool
+	evalers  []*idEvaler
+	env      *eval.Env
+	callback Callback
+}
+
+type variableItem struct {
+	Key    types.String
+	Values []types.Value
+}
+
+const unknownEntityType = "__cedar::unknown"
+
+func unknownEntity(v types.String) types.EntityUID {
+	return types.NewEntityUID(unknownEntityType, v)
+}
+
+var errUnboundVariable = fmt.Errorf("unbound variable")
+var errUnusedVariable = fmt.Errorf("unused variable")
+var errMissingPart = fmt.Errorf("missing part")
+var errInvalidPart = fmt.Errorf("invalid part")
+
+// Authorize will run a batch of authorization evaluations.
+//
+// All the request parts (PARC) must be specified, but you can
+// specify [Variable] or [Ignore].  Varibles can be enumerated
+// using the Variables.
+//
+// Using [Ignore] you can ask questions like "When ignoring context could this request be allowed?"
+//
+//  1. When a Permit Policy Condition refers to an ignored value, the Condition is dropped from the Policy.
+//  2. When a Forbid Policy Condition refers to an ignored value, the Policy is dropped.
+//  3. When a Scope clause refers to an ignored value, that scope clause is set to match any.
+//
+// Errors may be returned for a variety of reasons:
+//
+//   - It will error in case of a context error.
```

Comment:
This line needs some more words, methinks.

Reply:
Good catch, is this better?
```
//   - It will error in case of a context.Context error (e.g. cancellation).
```